keel-core/src/main/kotlin/com/netflix/spinnaker/keel/actuation/EnvironmentPromotionChecker.kt

```diff
@@ -141,14 +157,21 @@ class EnvironmentPromotionChecker(
                    * We don't need to re-invoke stateful constraint evaluators for these, but we still
                    * check stateless constraints to avoid approval outside of allowed-times.
                    */
+                  log.info("version: [$v] of artifact ${artifact.name} is in queuedForApproval for environment ${environment.name}")
                   if (checkStatelessConstraints(artifact, deliveryConfig, v, environment)) {
                     approveVersion(deliveryConfig, artifact, v, environment.name)
                     repository.deleteQueuedConstraintApproval(deliveryConfig.name, environment.name, v)
                   }
                 }
             }
             if (!approvedPending && versionIsPending || version == null) {
-              log.warn("No version of {} passes constraints for environment {}", artifact.name, environment.name)
+              if (version != null) {
+                val approvedVersion = repository.latestVersionApprovedIn(deliveryConfig, artifact, environment.name)
+                approvedVersion != null
+                  ?: log.info("version: [$approvedVersion] of artifact ${artifact.name} is currently approved for environment ${environment.name}")
```

Comment:
I'm confused by the `approvedVersion != null ?:` part of this expression.  My understanding is that the expression to the left of the `?:` operator has to be a nullable, but here it's a Boolean (`approvedVersion != null`). 

Are you trying to guard the log statement against `approvedVersion` being null?

Reply:
Yes :) I mean, if its null, the log statement will be `version null of xxx..` which is kinda weird in my opinion

Reply:
I don't think the code achieves that. The `?:` is really only for handling cases where there's a null. If you want the logic to fire only when it's non-null, I think `?.let` is a better choice:

```kotlin
approvedVersion?.let {approved -> log.info("version: [$approved] of artifact ${artifact.name} is currently approved for environment ${environment.name}")}
```

Reply:
I'm pretty sure you're only having to do this due to precedence. `?:` is applying to just the `null` keyword on line 170, not the `approvedVersion != null` expression (which is a boolean expression that would never evaluate to null).

You could rewrite this as:
```kotlin
approvedVersion?.also {
  log.info("version: [$approvedVersion] of artifact ${artifact.name} is currently approved for environment ${environment.name}")
}
```

Reply:
Couldn't it just be rewritten as an `if` statement? ```if (approvedVersion == null) {
  log.info("version: [$approvedVersion] of artifact ${artifact.name} is currently approved for environment ${environment.name}")
}```

Reply:
@robfletcher  It was written as an `if` statement and then I wanted to make it fancier ;) I will just add that `if` back.